src/uri/path.rs

```diff
@@ -19,74 +19,56 @@ const NONE: u16 = ::std::u16::MAX;
+                    b'?' if query.is_none() => Ok((Some(i), i+1)),
 
                     // This is the range of bytes that don't need to be
                     // percent-encoded in the path. If it should have been
                     // percent-encoded, then error.
+                    //
+                    // See https://url.spec.whatwg.org/#path-state
                     0x21 |
                     0x24..=0x3B |
                     0x3D |
                     0x40..=0x5F |
                     0x61..=0x7A |
                     0x7C |
-                    0x7E => {},
-
-                    _ => return Err(ErrorKind::InvalidUriChar.into()),
+                    0x7E => Ok((query, i+1)),
+
+                    // While queries *should* be percent-encoded, most
+                    // bytes are actually allowed...
+                    // See https://url.spec.whatwg.org/#query-state
+                    //
+                    // Allowed: 0x21 / 0x24 - 0x3B / 0x3D / 0x3F - 0x7E
+                    //
+                    // The list below is the bytes that are allowed in the
+                    // query but not in the path. The bytes allowed in both
+                    // are matched in the previous arm.
+                    0x3F |
+                    0x60 |
+                    0x7B |
+                    0x7D if query.is_some() => Ok((query, i+1)),
+
+                    // all bytes 0x80 and above match here (among others)
+                    // so all Ok() returns in the other match arms identify
+                    // a single byte UTF-8 code point
+                    _ => Err(InvalidUri(ErrorKind::InvalidUriChar)),
                 }
-            }
+            })?;
 
-            // query ...
-            if query != NONE {
-                for (i, &b) in iter {
-                    match b {
-                        // While queries *should* be percent-encoded, most
-                        // bytes are actually allowed...
-                        // See https://url.spec.whatwg.org/#query-state
-                        //
-                        // Allowed: 0x21 / 0x24 - 0x3B / 0x3D / 0x3F - 0x7E
-                        0x21 |
-                        0x24..=0x3B |
-                        0x3D |
-                        0x3F..=0x7E => {},
-
-                        b'#' => {
-                            fragment = Some(i);
-                            break;
-                        }
-
-                        _ => return Err(ErrorKind::InvalidUriChar.into()),
-                    }
-                }
-            }
-        }
-
-        if let Some(i) = fragment {
-            src.truncate(i);
-        }
+        src.truncate(len as usize);
 
         Ok(PathAndQuery {
+            // Safety: The try_fold() checks that each byte in the now truncated
+            // src is a single byte UTF-8 code point so src as a whole is valid
+            // UTF-8.
```

Comment:
Nit, take it or leave it: I might split this into two sentences. Currently, it feels like a bit of a run-on.

Maybe:
```suggestion
            // Safety: The try_fold() checks that each byte in the (now truncated)
            // src is a single byte UTF-8 code point. Therefore,  src as a whole
            // is valid UTF-8.
```
or something?

Reply:
Fixed.